pkg/dockerregistry/server/app.go

```diff
@@ -0,0 +1,143 @@
+	// it. In other cases it is hidden from us.
+	registry distribution.Namespace
+
+	// cachedLayers is a shared cache of blob digests to repositories that have previously been identified as
+	// containing that blob. Thread safe and reused by all middleware layers. It contains two kinds of
+	// associations:
+	//  1. <blobdigest> <-> <registry>/<namespace>/<name>
+	//  2. <blobdigest> <-> <namespace>/<name>
+	// The first associates a blob with a remote repository. Such an entry is set and used by pullthrough
+	// middleware. The second associates a blob with a local repository. Such a blob is expected to reside on
+	// local storage. It's set and used by blobDescriptorService middleware.
+	cachedLayers digestToRepositoryCache
+
+	// quotaEnforcing contains shared caches of quota objects keyed by project
+	// name. Will be initialized only if the quota is enforced.
+	// See EnforceQuotaEnvVar.
+	quotaEnforcing *quotaEnforcingConfig
+}
+
+// NewApp configures the registry application and returns http.Handler for it.
+// The program will be terminated if an error happens.
+func NewApp(ctx context.Context, registryClient client.RegistryClient, dockerConfig *configuration.Configuration, extraConfig *registryconfig.Configuration, writeLimiter maxconnections.Limiter) http.Handler {
+	app := &App{
+		ctx:            ctx,
+		registryClient: registryClient,
+		extraConfig:    extraConfig,
+		writeLimiter:   writeLimiter,
+	}
+
+	cache, err := newDigestToRepositoryCache(defaultDigestToRepositoryCacheSize)
+	if err != nil {
+		panic(err)
+	}
+	app.cachedLayers = cache
+
+	weaveAppIntoConfig(app, dockerConfig)
+
+	for _, middleware := range dockerConfig.Middleware["repository"] {
+		if middleware.Name == middlewareOpenShift {
+			rc, err := newRepositoryConfig(ctx, middleware.Options)
+			if err != nil {
+				context.GetLogger(ctx).Fatalf("error configuring the repository middleware: %s", err)
+			}
+			app.repositoryConfig = rc
+			app.quotaEnforcing = newQuotaEnforcingConfig(ctx, os.Getenv(EnforceQuotaEnvVar), os.Getenv(ProjectCacheTTLEnvVar), middleware.Options)
+			break
+		}
+	}
+
+	dockerApp := handlers.NewApp(ctx, dockerConfig)
+
+	// Add a token handling endpoint
+	if dockerConfig.Auth.Type() == middlewareOpenShift {
+		tokenRealm, err := TokenRealm(dockerConfig.Auth[middlewareOpenShift])
+		if err != nil {
+			context.GetLogger(dockerApp).Fatalf("error setting up token auth: %s", err)
+		}
+		err = dockerApp.NewRoute().Methods("GET").PathPrefix(tokenRealm.Path).Handler(NewTokenHandler(ctx, registryClient)).GetError()
+		if err != nil {
+			context.GetLogger(dockerApp).Fatalf("error setting up token endpoint at %q: %v", tokenRealm.Path, err)
+		}
+		context.GetLogger(dockerApp).Debugf("configured token endpoint at %q", tokenRealm.String())
+	}
+
+	app.registerBlobHandler(dockerApp)
+
+	// Registry extensions endpoint provides extra functionality to handle the image
+	// signatures.
+	RegisterSignatureHandler(dockerApp)
```

Comment:
*nit/joke*: I need to remember to register next special handler using locally defined closure or something more creative.